components/org.wso2.carbon.identity.application.authenticator.biometric/src/main/java/org/wso2/carbon/identity/application/authenticator/biometric/BiometricAuthenticator.java

```diff
@@ -0,0 +1,152 @@
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Biometric Authenticator class.
+ */
+public class BiometricAuthenticator extends AbstractApplicationAuthenticator
+        implements FederatedApplicationAuthenticator {
+
+    private static final Log log = LogFactory.getLog(BiometricAuthenticator.class);
+    private static String sessionDataKey;
+    private static String randomChallenge;
+    private static String signedChallenge;
+
+    /**
+     * Get the friendly name of the Authenticator.
+     */
+    @Override
+    public String getFriendlyName() {
+
+        return BiometricAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+
+        signedChallenge = request.getParameter(BiometricAuthenticatorConstants.SIGNED_CHALLENGE);
+        return signedChallenge != null;
+    }
+
+    @Override
+    public String getContextIdentifier(javax.servlet.http.HttpServletRequest request) {
+
+        sessionDataKey = request.getParameter(BiometricAuthenticatorConstants.SESSION_DATA_KEY);
+        return sessionDataKey;
+    }
+
+    /**
+     * Get the name of the Authenticator.
+     */
+    @Override
+    public String getName() {
+
+        return BiometricAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
+                                                 AuthenticationContext context) {
+
+        AuthenticatedUser user = context.getSequenceConfig().getStepMap().get(1).getAuthenticatedUser();
+        String username = user.getUserName();
+
+        Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+        String serverKey = authenticatorProperties.get(BiometricAuthenticatorConstants.SERVER_KEY);
+
+        String hostname = IdentityUtil.getHostName();
+        String serviceProviderName = context.getServiceProviderName();
+        String message = username + " is trying to log into " + serviceProviderName + " from " + hostname;
+
+        UUID challenge = UUID.randomUUID();
+        randomChallenge = challenge.toString();
+
+        BiometricDAOImpl biometricDAO = BiometricDAOImpl.getInstance();
+        String deviceID = biometricDAO.getDeviceID(username);
+
+        PushNotificationSenderImpl pushNotificationSender = PushNotificationSenderImpl.getInstance();
+        pushNotificationSender.sendPushNotification(deviceID, serverKey, message, randomChallenge, sessionDataKey);
+
+        try {
+            String pollingEndpoint = "https://biometricauthenticator.private.wso2.com:9443/" +
```

Comment:
Use framework utils to append the query params to the url and we need to check the posibiity of reading this host name from file or configuration